### PR TITLE
Stop loading the entire node set into memory per tsh ssh connection

### PR DESCRIPTION
**Pull request by rosstimothy**
When doing a `tsh ssh node` the proxySubsys would load the entire
set of nodes into memory to determine which server to route the
request to. Under heavy load, like an automated bot that periodically
spawns numerous session, a proxy could easily consume all
available memory.

To do prevent this, proxies now utilize a NodeWatcher, that maintains
a single node set in memory. This prevents loading the nodes into
memroy more than once, and also eliminates the need to unmarshal the
types.Server on each retrieval of the nodes. The NodeWatcher only
provides a GetNodes function that require a filter function to make
it intentionally challenging to retrieve a copy of the entire node set.

### CODE REVIEW

**Comment by rosstimothy** [comment]
friendly ping @smallinsky @ibeckermayer

File: lib/srv/regular/sshserver.go [hunk 1]
```diff
@@ -187,6 +187,9 @@ type Server struct {
 
 	// lockWatcher is the server's lock watcher.
 	lockWatcher *services.LockWatcher
+
+	//nodeWatcher is the server's node watcher.
```
Comment (smallinsky) [inline lib/srv/regular/sshserver.go:191]:
```suggestion
	// nodeWatcher is the server's node watcher.
```

**Review by smallinsky — COMMENTED** [review]
@rosstimothy
Sorry for the delay. I have left one question about `NodeWatcher` approach vs access point cache where the NodeType is already watched 
https://github.com/gravitational/teleport/blob/1aa38f4bc56997ba13b26a1ef1b4da7a3a078930/lib/cache/collections.go#L691

**Review by smallinsky — APPROVED** [review]
> By leveraging the NodeWatcher we can be sure that the entire node set is only loaded into memory once, and by storing them as types.Server we don't have to pay the cost to unmarshal from JSON. The method by which a caller retrieves nodes from the NodeWatcher is also meant to ensure that only copies of the matched subset are made.

All right, thanks for the explanation. 

I was curios what is actually benefit and checked  benchmark for old cache approach vs watcher for 40_000 nodes: 
```
BenchmarkListNodesWatcher-8   	      16	  68339640 ns/op	29017941 B/op	   40183 allocs/op
BenchmarkListMaxNodes-8   	           5      229637793 ns/op	80429200 B/op	 1520193 allocs/op
```

I have left some minor comment otherwise  it LGTM.

**Review by ibeckermayer — APPROVED** [review]
@rosstimothy apologies, I've been head down and wasn't giving needed attention to my GitHub notifications. 

Possibly a naive question and/or the premise may be false, but aren't we using the `GetNodes` you're replacing in many other parts of the codebase? Why should those not also be replaced with a call to a `NodeWatcher` to prevent a similar potential problem?

**Comment by rosstimothy** [comment]
> @rosstimothy apologies, I've been head down and wasn't giving needed attention to my GitHub notifications.
> 
> Possibly a naive question and/or the premise may be false, but aren't we using the `GetNodes` you're replacing in many other parts of the codebase? Why should those not also be replaced with a call to a `NodeWatcher` to prevent a similar potential problem?

Good question. Anything service other than auth which could potentially call `GetNodes` concurrently should definitely make use of the `NodeWatcher`. Doing a quick check it looks like there may be a couple spots in the web api that should probably be migrated.

**Pull request merged**